apps/drupal/src/Plugin/Preprocessor/PreprocessorManager.php

```diff
@@ -0,0 +1,509 @@
+<?php
+
+namespace Drupal\particle\Plugin\Preprocessor;
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\particle\Tools\ParticleTools;
+
+/**
+ * Provides a theme preprocessor manager for Class based theme preprocessors.
+ *
+ * Drupal core does not provide built-in plugin support for theme files.
+ * The plugin discovery might before the front end theme is not active.
+ *
+ * The approach here is to use Drupal's State API to handle Preprocessor Class.
+ * New suggestions are read on page load and added to State API via STATE_ID.
+ * The Preprocessor Class Map is stored in cache and refreshed on change.
+ *
+ */
+class PreprocessorManager {
+
+  /**
+   * The State API ID for this Manager.
+   *
+   * @var string
+   */
+  const STATE_ID = 'preprocessor.map';
+
+  /**
+   * Appended suggestions.
+   *
+   * @see getSuggestions()
+   *
+   * @var array
+   */
+  protected static $allSuggestions = [];
+
+  /**
+   * Cached Preprocessor instances.
+   *
+   * @see load()
+   *
+   * @var array
+   */
+  protected static $preprocessorInstances = [];
+
+  /**
+   * Particle tools instance for all preprocessors.
+   *
+   * @var Drupal\particle\Tools\ParticleTools
+   *   Particle's suite of helper functions and utilities.
+   */
+  protected static $tools;
+
+  /**
+   * Returns the tools object.
+   *
+   * @return Drupal\particle\Tools\ParticleTools
+   *   The tools object.
+   */
+  public static function getTools() {
+    if (!isset(static::$tools)) {
+      static::$tools = new ParticleTools();
+    }
+
+    return static::$tools;
+  }
+
+  /**
+   * Returns the full path to the directory of the preprocessor classes.
+   *
+   * Base should be the only root level preprocessor in this directory.
+   * All other types should be located in subdirectories.
+   *
+   * @see getPreprocessors()
+   *
+   * @return string
+   *   The directory path.
+   */
+  public static function getPreprocessorsDirectory() {
+    return __DIR__ . '/Preprocessors';
+  }
+
+  /**
+   * Get the preprocessors from directory and subdirectories.
+   *
+   * @return array
+   *   An array of preprocessors from the file system.
+   */
+  public static function getPreprocessors() {
+    $preprocessors = &drupal_static(__FUNCTION__);
+    if (is_null($preprocessors)) {
+      // Get all files in the preprocessors directory.
+      $files = file_scan_directory(static::getPreprocessorsDirectory(), '/.php/');
```

Comment:
I wonder if these can be further cached? I'd be worried about potential performance of doing a file scan on each page load...

Reply:
Since this is only ever called in mergedMaps, which is behind a check for new suggestions, I think it's okay. This should only run when there's something to update, which seems appropriate to me. But absolutely agree a better caching scheme might be good! But I think we'll need some evidence from actually using this on a project to move forward.

Reply:
I think this is another area where the normal Cache API can be used.  A file scan in a targeted directory like this might not seem bad, but I've seen several client instances with slow file systems where any sort of file access needed to be avoided and cached.  I feel like the actual performance is less of an issue than the architecture of this.  And again it's a perfect use-case for a cache bin.